packages/base/src/mainview/mainview.tsx

```diff
@@ -97,6 +103,246 @@ export class MainView extends React.Component<IProps, IStates> {
     }
   };
 
+  /**
+   * MapLibre function to execute operation on `style` (add/update/remove layer/source),
+   * avoiding "Map.style undefined" error.
+   * This is required because of the lack of 'ready' promise in the Map object.
+   *
+   * @param callback - the function updating the Map.
+   */
+  private _mapLibreExecute(callback: () => void) {
+    // Workaround to avoid "Map.style undefined" error, because of the miss of a
+    // 'ready' promise.
+    if (this._Map.loaded()) {
+      callback();
+    } else {
+      this._Map.on('load', callback);
+    }
+  }
+
+  /**
+   * Add a source in the map.
+   *
+   * @param id - the source id.
+   * @param source - the source object.
+   */
+  addSource(id: string, source: IJGISSource): void {
```

Comment:
Should we start extracting these APIs in their own class implementing an `IMapViewer` interface? 

That way we prepare the ground for future possible implementations/plugins